### PR TITLE
Pull Request to Add Local and Component-wise Prediction Rigidity

### PR DESCRIPTION
**Pull request by SanggyuChong**
Hello all,

This PR aims to introduce new features to `scikit-matter`, which allows the user to calculate what is called the local prediction rigidity (LPR) and the (local) component-wise prediction rigidity ((L)CPR).

LPR and CPR are metrics that allow one to assess the robustness, or quite literally the "rigidity", in the local or component-wise prediction that the machine learning model makes. More details can be found in this preprint: https://arxiv.org/pdf/2306.15638.pdf
(Note that CPR is not explained in the above preprint. It is, however, a metric that follows the same spirit as the LPR, and is something that will be shared in forthcoming works by myself and coworkers.)

I found these things to best fit in with the `metrics`, and hence the functions have been coded in there. Key points to note, among other things, is that the function assumes the user to (1) provide the regularization strength of a model that has been trained, and (2) consider the LPR and CPR only in the contexts where training is done on _**structural average**_ feature vectors and targets, to keep the resulting metric within an interpretable and roughly transferable range.

Please let me know what you think. Tagging and assigning @agoscinski as he has helped me write up this PR. Will only post as draft at the moment.

### CODE REVIEW

File: src/skmatter/metrics/_prediction_rigidities.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    comp_idxs = np.cumsum([0] + comp_dims.tolist())
+    X_test = np.vstack(X_test)
+    atom_count = X_test.shape[0]
+    CPR = np.zeros((num_test, num_comp))
+    LCPR_np = np.zeros((atom_count, num_comp))
+
+    for ci in range(num_comp):
+
+        tot_comp_idx = np.arange(comp_dims.sum()) 
+        mask = ((tot_comp_idx >= comp_idxs[ci]) & (tot_comp_idx < comp_idxs[ci+1])).astype(float)
```
Comment (agoscinski) [inline src/skmatter/metrics/_prediction_rigidities.py:184]: It seems to me that you just take ranges of the matrix in the end. Why don't you just do?
```python
comp_dims = [0] + comp_dims.tolist()
#...
  A = ... # some calcluators
  Xic = A[:, comp_dims[i]:comp_dims[i+1]]
```
I dont know if I am missing something, but at the moment it seems to me a convoluted way to do this.

**Review by agoscinski — COMMENTED** [review]
Can you add tests into the `test/test_metrics.py`? You can take the `ReconstructionMeasuresTests` class as refernce

File: src/skmatter/preprocessing/_data.py [hunk 2]
```diff
@@ -154,17 +154,16 @@ def fit(self, X, y=None, sample_weight=None):
...
         self.scale_ = 1.0
-        if self.with_std:
-            X_mean = np.average(X, weights=sample_weight, axis=0)
-            var = np.average((X - X_mean) ** 2, weights=sample_weight, axis=0)
 
+        if self.with_std:
+            var = np.average((X - self.mean_) ** 2, weights=sample_weight, axis=0)
             if self.column_wise:
-                if np.any(var < self.atol + abs(X_mean) * self.rtol):
+                if np.any(var < self.atol + abs(self.mean_) * self.rtol):
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/skmatter/preprocessing/_data.py:161]: I checked with scikit-learn: The way it was before is actually conform wtih sklearn's StandardScaler which makes sense also to me because the parameter is called `with_std` so standard deviation includes the mean. Can you do it just manually for your code?

File: src/skmatter/metrics/_prediction_rigidities.py (hunk 1 above)
Comment (SanggyuChong) [reply src/skmatter/metrics/_prediction_rigidities.py:184]: Thanks for the catch. you are right, i was mindlessly coding and had end up with something convoluted.

File: src/skmatter/metrics/_prediction_rigidities.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    X_struc = []
+    for i, cur_X in enumerate(X_train):
+        X_struc.append(np.mean(scaler.transform(cur_X), axis=0))
+    X_struc = np.vstack(X_struc)
+
+    # build XX and obtain Xinv for LPR calculation
+    XX = X_struc.T @ X_struc
+    Xprime = XX + reg * np.eye(XX.shape[0])
+    rank_diff = X_struc.shape[1] - np.linalg.matrix_rank(Xprime)
+    Xinv = np.linalg.inv(Xprime)
```
[the remark this answers is not in the document]
Comment (SanggyuChong) [reply src/skmatter/metrics/_prediction_rigidities.py:67]: OK, will replace with pinv for now for improved numerical stability. This could be helpful especially in cases where the user, for example, wants a purely linear model w/o regularization. Need to experiment with this further.

File: src/skmatter/metrics/_prediction_rigidities.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    X_test : list of ndarray of shape (n_atoms, n_features)
+        Test dataset where each training set structure is stored as a separate 
+        ndarray.        
+
+    reg : float value
+        Regularizer value that the linear/kernel model has been optimized to.
+
+    comp_dims : array of int values
+        Dimensions of the feature vectors pertaining to each prediction component.
```
[the remark this answers is not in the document]
Comment (SanggyuChong) [reply src/skmatter/metrics/_prediction_rigidities.py:132]: no, i meant "pertaining", like the word

File: src/skmatter/metrics/_prediction_rigidities.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+
+    The function assumes that the feature vectors for the local environments and 
+    structures are built by concatenating the descriptors of different prediction
+    components together. It also assumes, like the case of LPR, that model training
+    is undertaken in a manner where the global prediction targets are averaged over
+    the number of atoms appearing in each training structure, and the average feature
+    vector of each structure is hence used in the regression. Likewise, to guarantee
+    valid comparison in the (L)CPR across different models, feature vectors are scaled
+    by a global factor based on standard deviation across atomic envs.
```
[the remark this answers is not in the document]
Comment (SanggyuChong) [reply src/skmatter/metrics/_prediction_rigidities.py:116]: I agree that in theory, this can be applied outside of atomistic learning. Yet, I find it hard to explain the LPR and CPR in a purely theoretical sense. On top of that, as this has been written for scikit-_matter_, I think we should be fine just keeping the current doc that focuses its explanations on atomistic learning.

File: src/skmatter/preprocessing/_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SanggyuChong) [reply src/skmatter/preprocessing/_data.py:161]: Ok, will do. This scaling choice was a heuristic one, where we found the scaling to pseudo standard deviation (compute by forcing the mean at **0**) to be the most numerically stable in the analyses that we were performing.

File: src/skmatter/metrics/_prediction_rigidities.py (hunk 1 above)
Comment (SanggyuChong) [reply src/skmatter/metrics/_prediction_rigidities.py:184]: Thanks for the catch. Here however, we must keep the original feature vector length. I will come up with some middle ground..

File: src/skmatter/metrics/_prediction_rigidities.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/skmatter/metrics/_prediction_rigidities.py:132]: Then I am not sure what exactly is meant, it is nowhere else mentioned what pretraining is

**Review by agoscinski — COMMENTED** [review]
thanks looks good, just had minor comments. Can u also add your code doc to the sphinx doc? You just need to add the prediction rigidity to this file  https://github.com/scikit-learn-contrib/scikit-matter/blob/main/docs/src/references/metrics.rst using autofunction as for the reconstruction measures. The text of the automodule can be changed in this file https://github.com/scikit-learn-contrib/scikit-matter/blob/main/src/skmatter/metrics/__init__.py At the moment it only mentions the reconstructions measures. Feel free to change the text so it includes better both measures and ask questions, sphinx doc is a bit confusing and frustrating at the beinning.

File: tox.ini [hunk 6]
```diff
@@ -81,3 +81,4 @@ exclude =
 per-file-ignores =
     # D205 and D400 are incompatible with the requirements of sphinx-gallery
     examples/**:D205, D400
+extend-ignore = E203
```
[the remark this answers is not in the document]
Comment (SanggyuChong) [reply tox.ini:84]: this should be revisited. for now, i will just make it specific to my file.

**Review by agoscinski — APPROVED** [review]
Looks good!
